### ISSUE TITLE
how to import server side HotColumn such as using nextjs?

### ISSUE THREAD

**Issue by edmodelorg** [handsontable/handsontable#7445]
### Steps to reproduce

1.
2.
3.

### Demo

https://jsfiddle.net/handsoncode/8ffpsqt6/

### Your environment
* Handsontable version: latest
* Browser Name and version: latest
* Operating System: latest

**Comment by AMBudnik** [comment]
Hi @edmodelorg 

Can you provide some more details about running up the project? I've never tested nextJS.
Do you get any errors? Are you able to display anything?

**Comment by AMBudnik** [comment]
Hey @edmodelorg 

I have an update on this. Here's a related topic https://github.com/handsontable/vue-handsontable-official/issues/70
Especially the last comment from @wojciechczerniak can help us to solve this case

> Missing navigator is a configuration issue with SSR. Please follow this article: https://medium.com/@codebeast_/why-your-third-party-plugin-dont-work-in-nuxt-and-how-to-fix-it-d1a8caadf422 to configure SSR for Handsontable

**Comment by edmodelorg** [comment]
Hi @AMBudnik 

Sorry for the late reply. I am using react nextjs not vue nuxtjs. I was trying to incorporate similar vendor plugins to next.config.js, however I came across nextjs site indicating that next.config.js will not be used during browser build, so any similar solutions perhaps will not work.

There was no issue rendering HotTable using the following import.

const HotTable = dynamic(
  () => import('@handsontable/react').then(mod=>{
    return mod;
  }),
  { ssr: false }
);

However I got windows not defined error while trying to import Handsontable or HotColumn. 

//This will throw windows not defined error
import { HotTable, HotColumn } from "@handsontable/react";

when I switch to use the dynamic import for HotColumn, the column does not render at all.

//This import also doesn't find the handsontable object
const Handsontable = dynamic(() =>
  import('handsontable'),
  {ssr: false}
);

//This import does not render
const { HotColumn } = dynamic(() =>
  import('@handsontable/react'),
  {ssr: false}
);

The following code does not render

  <HotColumn data='firstColumn'>
              <MyCustomColumnRenderer hot-renderer/>
  </HotColumn>

**Comment by swistach** [comment]
Hi @edmodelorg,
Here is a simple demo with `next/dynamic` with disabled `ssr`: https://codesandbox.io/s/serene-ardinghelli-0z1zb
I used an optional `loading` component (https://nextjs.org/docs/advanced-features/dynamic-import#with-custom-loading-component). It works but I'd say it's a workaround. The problem lies in `handsontable/helpers/feature`: https://github.com/handsontable/handsontable/blob/master/src/helpers/feature.js#L5
`window` is undefined in server-side.

@wojciechczerniak - maybe we should do a similar fix as we did with using `global.console`? It definitely needs some additional investigation in a separated project.

**Comment by Surecom** [comment]
@swistach it would be great, if handsontable team be able to fix it, and allow SSR for this library - a significant part of the community will say thank to you :)

**Comment by gscoon** [comment]
I found the approach below to be significantly simpler IMO.  I simply separated the dynamic import code from the HotTable code by doing the following:

1. Create a page file `pages/mypage.js`
2. Create a separate component file `components/MyCustomTable.jsx`

In `mypage.js`:
```
import React from 'react';
import dynamic from 'next/dynamic';

const MyCustomTable = dynamic(
    () => import('./../components/MyCustomTable.jsx'),
    { ssr: false }
);

export default function MyPage() {
    return (
        <MyCustomTable />
    );
}
```

in `MyCustomTable.jsx`
```
import React from 'react';
import { HotTable } from '@handsontable/react';
import HandsonTable from 'handsontable';

export default class MyCustomTable extends React.Component {
    tableRef = React.createRef();

    componentDidMount() {
        console.log('Table ref', this.tableRef.current)
    }

    render() {
        return (
            <HotTable
                ref={this.tableRef}
                // other table props go here
            />
        );
    }
}
```

This solved 2 problems for me:
- Avoiding the dreaded "window" undefined error
- Using a ref (React.createRef) to access the table instance

**Comment by swistach** [comment]
We classified this issue as a bug due to the increasing usage of SSR.
Before we start working on this bug, we suggest using the above workarounds.

**Comment by wojciechczerniak** [comment]
@swistach I wouldn't call it a workaround, it's the correct way to handle client side only components. And would not classify this as a bug, making client side only component supported in SSR is an improvement suggestion. It was an assumption that Handsontable will always run within the web browser environment and no such feature was ever implemented to consider this as a bug.

**Comment by swistach** [comment]
In my opinion, our use of global `window` is a bug, and most likely, it breaks SSR not only for Nextjs. But I wouldn't like to haggle about it. I change the label to `improvement suggestion`. Thank you for your opinion!

**Comment by gka** [comment]
> In my opinion, our use of global window is a bug, and most likely, it breaks SSR not only for Nextjs.

Yep, this also breaks SSR in Svelte. Would be really good to remove that access, or at least wait until the table is being initialized. This way one could at least load the HOT code during server-side rendering but postpone rendering until CSR.

**Comment by aboveyunhai** [comment]
in the latest `next 13`, I'm not sure about the `12`,
every single import from `handsontable` regardless the import item is a `component` or `general api/helpers` will raise `window-related api` error due the to code was evaluated in the server side.

```js
\\ the following import will be all broken
\\ For example: 

import Handsontable from "handsontable", "registry" , "renderer", etc. //error

Handontable.whatever // error
```
which means in order to use handontable in `next.js`. You pretty much have to hack through everything via some weird client side and loading inside `useEffect()`. 

While the `next/dynamic` or `react lazy load` can solve the basic HotTable component usage, 
note that the components like 
```
const { HotColumn } = dynamic(() => import('@handsontable/react'), {ssr: false} );
``` 
will not render at all (The HotColumn would be executed for some reason). 
That means you will completely give up the advantage of using `custom renderer component` and have to use the custom renderer functions, which was discouraged from the official doc:
https://handsontable.com/docs/react-data-grid/cell-renderer/#declare-a-custom-renderer-as-a-component

**Comment by 4F2E4A2E** [comment]
I am evaluating ag-grid-react and this looks a bit like a deal breaker to us. We would like to have full support of next13, ssr and react-server-components: https://www.joshwcomeau.com/react/server-components/

**Comment by adrianszymanski89** [comment]
Hi @4F2E4A2E 

Thank you for your input. We have an internal issue reported for further research regarding SSR. We will update this thread once we have more information about the possible solutions and future plans.

**Comment by craigcartmell** [comment]
Hey @adrianszymanski89 - we're also comparing a few different data grids here, and unfortunately Handsobtable is a none starter for us currently due to the lack of SSR support, or a viable workaround.

Is there any progress on your side for this?

**Comment by adrianszymanski89** [comment]
Hi @craigcartmell 

Just recently, we started the investigation process on SSR support with Handsontable. We need some time to prepare the specifications and check the possible solutions, so I'll update you once we have more info.

**Comment by evanSe** [comment]
Hey @craigcartmell a workaround for now would be to use a dynamic import with `ssr: false`

```jsx
// components/Table.tsx
import { HotTable } from "@handsontable/react";
import "handsontable/dist/handsontable.full.min.css";


export default function Table( ) {
  return (
    <div className=" flex h-full flex-grow flex-col bg-[#442525]">
      <HotTable
        contextMenu
        allowInsertColumn
        allowInsertRow
        rowHeaders={true}
        colHeaders={true}
        // ...{rest of your settings}
      />
    </div>
  );
}


// index.tsx
const Table = dynamic(
  () => import("~/components/Table"),
  {
    ssr: false,
  }
);

export default function Home() {
  return (
    <Table />
  );
}
```

**Comment by craigcartmell** [comment]
Hey @evanSe  - I seem to have the table component working with a similar workaround, but I can't get hotcolumns working.

Do you have an example of how to import `<HotColumn>` without ssr?

**Comment by evanSe** [comment]
> Hey @evanSe - I seem to have the table component working with a similar workaround, but I can't get hotcolumns working.
> 
> Do you have an example of how to import `<HotColumn>` without ssr?
> 
> Sounds like I'm facing the same issue as here: [#7445 (comment)](https://github.com/handsontable/handsontable/issues/7445#issuecomment-1497004307)

Hey @craigcartmell sorry for not getting back sooner, GitHub notifications can be a mess!

You can use HotColumn if you include it inside the Table.tsx file, as shown in the following example. In the next version of Handsontable, we will be making some improvements to our support for Next.js, which means you won't need to perform a dynamic import. Instead, simply add a 'use client' statement at the top of your component. For now, you can refer to the following example.

```
// index.tsx
import dynamic from "next/dynamic";

const Table = dynamic(() => import("~/components/Table"), {
  ssr: false,
});

export default function Home() {
  return <Table />;
}

// Table.tsx
import React from "react";
import { HotColumn, HotTable } from "@handsontable/react";
import "handsontable/dist/handsontable.full.min.css";

const RendererComponent = (props: { row?: number; value?: string; col?: number }) => {
  return (
    <>
      <i style={{ color: "#a9a9a9" }}>
        Row: {props.row}, column: {props.col},
      </i>{" "}
      value: {props.value}
    </>
  );
};

const hotData = [
  ["A1", "B1", "C1", "D1", "E1"],
  ["A2", "B2", "C2", "D2", "E2"],
  ["A3", "B3", "C3", "D3", "E3"],
  ["A4", "B4", "C4", "D4", "E4"],
  ["A5", "B5", "C5", "D5", "E5"],
  ["A6", "B6", "C6", "D6", "E6"],
  ["A7", "B7", "C7", "D7", "E7"],
  ["A8", "B8", "C8", "D8", "E8"],
  ["A9", "B9", "C9", "D9", "E9"],
];

export default function Table() {
  return (
    <div className=" flex h-full flex-grow flex-col bg-[#442525]">
      <HotTable data={hotData} licenseKey="non-commercial-and-evaluation">
        <HotColumn width={250}>
          {/* add the `hot-renderer` attribute to mark the component as a Handsontable renderer */}
          <RendererComponent hot-renderer />
        </HotColumn>
      </HotTable>
    </div>
  );
}
```

**Comment by sl01k** [comment]
This has been addressed in PR #[10575](https://github.com/handsontable/handsontable/pull/10575), which added full support for SSR/fullstack frameworks including Next.js, Remix, Gatsby, Nuxt, and SvelteKit.

Handsontable can now be used with a simple 'use client' directive in Next.js (both with next/dynamic and static imports), without the window is not defined errors.

Closing as resolved.
